### DISCUSSION TITLE
RF + XGBoost ensemble model for industrial plant

### DISCUSSION THREAD

**Discussion by emiliomastriani** [shap/shap#4071] [Q&A]
Hello everyone, my project is focused on predictive maintenance of an industrial plant. This means we are working with a huge collection of data from many sensors, which have been collected over a couple of years, and they rarely present anomalies.  We got an AUC-ROC near to 98% using the segmentation on the dataset, and applying separately the RF and the XGBoost to the dataset. Then we performed ensemble learning by combining the predictions of two machine learning models (a Random Forest and an XGBoost classifier) using averaged probabilities. This ensemble model was demonstrated to be the most valid in our scenario. Here is the question: now I need to provide the explainability of my hybrid model, can SHAP be used in this context? How must I proceed considering my ensemble model? Thank you all

**Comment by DanteTrb** [comment]
Thank you for raising this important question.. your use case (ensemble of RF + XGBoost for predictive maintenance with high AUC-ROC) is compelling and deserves a rigorous explainability approach.

✅ Yes, SHAP can be used on your ensemble, with care. While SHAP was originally designed to explain individual models, you can extend it to ensemble settings using one of the following strategies:

###  **Strategy 1: Explaining each model separately, then aggregating SHAP values**
Since both Random Forest and XGBoost are tree-based, you can use TreeExplainer for each:

```
import shap

explainer_rf = shap.TreeExplainer(rf_model)
explainer_xgb = shap.TreeExplainer(xgb_model)

shap_values_rf = explainer_rf.shap_values(X)
shap_values_xgb = explainer_xgb.shap_values(X)

# Weighted average if ensemble uses averaged probabilities
shap_values_ensemble = 0.5 * (shap_values_rf + shap_values_xgb)
```

🔍 If your ensemble uses non-uniform weights, adjust accordingly: α * SHAP_RF + (1–α) * SHAP_XGB

This approach gives a meaningful **approximation of the ensemble’s SHAP values**, preserving the local contributions from both learners.

### **Strategy 2: Training a surrogate model to mimic the ensemble**
Train a simple, interpretable model (e.g. shallow XGBoost or Logistic Regression) to approximate the ensemble’s outputs:

```
from sklearn.linear_model import LogisticRegression

ensemble_preds = 0.5 * (rf_model.predict_proba(X)[:, 1] + xgb_model.predict_proba(X)[:, 1])
surrogate = LogisticRegression().fit(X, ensemble_preds > 0.5)

explainer = shap.Explainer(surrogate)
shap_values = explainer(X)
```

✅ This provides a **proxy explanation** for the ensemble, but might lose fidelity depending on surrogate accuracy.

⚠️ Technical Notes:

SHAP assumes feature independence or low multicollinearity. Since you’re using tree models, this assumption is more acceptable.

However, if you suspect high multicollinearity, consider grouping correlated features or using SHAP interaction values for deeper insights.

Yes, you **can and should** use SHAP for your ensemble model. The best practice is to compute SHAP values on each base model and average them according to your ensemble logic. Alternatively, fit a surrogate model if interpretability over fidelity is acceptable. Both approaches are valid — choose based on your application’s needs.

Let me know if you’d like a working example or follow-up! If this helped, feel free to mark this as accepted so others can benefit too 🚀

**Reply by vikram-raju** [reply]
You cannot simply weigh the two models. Each model's SHAP values are relative to its respective model's expected value. So Feature_1 in model 1 with +0.15 does not necessarily mean it's more important than Feature 2 from model 2 with a shap score of +0.1 or +0.2. Since the deviation from baseline (0.35 vs 0.65) can be different. You're weighted averaging deviations from different reference points.
